File: Leetcode/Python/_1372.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(node, left: bool = True, steps: int = 0):
            if node:
                self.length_path = max(self.length_path, steps) 
                if left:
                    dfs(node.right, False, steps + 1)
                    dfs(node.left, True, 1)
                else:
                    dfs(node.left, True, steps + 1)
                    dfs(node.right, False, 1)
                    
                    
        self.length_path = 0
        dfs(root, True, 0)
        dfs(root, False, 0)
        return self.length_path

File: Leetcode/cpp/1372.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int path_length = 0;
    void dfs(TreeNode* node, bool left, int steps) {
        if (node == nullptr) {
            return;
        }
        path_length = max(path_length, steps);
        if (left) {
            dfs(node->right, false, steps + 1);
            dfs(node->left, true, 1);
        } else {
            dfs(node->left, true, steps + 1);
            dfs(node->right, false, 1);
        }
    }
    int longestZigZag(TreeNode* root) {
        dfs(root, true, 0);
        dfs(root, false, 0);
        return path_length;
    }
};
